server/aesdsocket.c

```diff
@@ -71,199 +111,323 @@ int daemon_creation()
-  			syslog(LOG_ERR, "Error: partial write with code, %d", errno);
-  		}
-  		
-  		data_count += write_return;
-  			
-  		memset(buffer, '\0', SIZE);
-  		lseek(socket_file_fd, 0, SEEK_SET);
-  		  			
-  		char *buffer_read = (char *)malloc(data_count);
-  		
-  		if (buffer_read == NULL)
-  		{
-            		syslog(LOG_ERR, "malloc");
-        	}
-  			
-  		int read_return = read(socket_file_fd, buffer_read, data_count);
-  		
-  		if (read_return == -1)
-  		{
-            		syslog(LOG_ERR, "Error: read() with code, %d", errno);
-  		}
-  			
-  		int send_return = send(client_socket_fd, buffer_read, data_count, 0);
-  		
-  		
-      		if (send_return == -1)
-      		{
-            		syslog(LOG_ERR, "Error: write() with code, %d", errno);
-        	}
-
-  		
-  		free(buffer_read);
-  		free(total_buffer);	
-  		syslog(LOG_DEBUG, "Closed connection from %s \n", ip);
-	}
 	
-	return 0;
+    socket_file_fd = open( "/var/tmp/aesdsocketdata", O_RDWR | O_CREAT | O_APPEND, 0666);
+
+
+
+    ret = timer_init();
+    if(ret == -1)
+    {
+        perror("timer creation");
+    }
+
+    pthread_mutex_init(&mtx, NULL);
+
+    node_t *head =NULL;
+    node_t *prev,*current;
+
+    while(!e_status)
+    {
+        if(timeout)
+        {
+            timeout = false;
+            timestamp();
+        }
+        int client_socket_fd  = accept(server_socket_fd, (struct sockaddr *)&their_addr, &addr_size);
+        if( client_socket_fd  == -1 )
+        {
+            if(errno == EWOULDBLOCK)
+            {
+          
+                continue;
+            }
+            syslog(LOG_ERR, "Error : accept with error no : %d", errno);
+      
+            continue;
+        }
+        node_t *new_node = (node_t *)malloc(sizeof(node_t));
+        new_node->data.complete_flag = false;
+        new_node->data.fd = client_socket_fd;
+        new_node->data.t_addr = their_addr;
+     
+        int ret = pthread_create( &(new_node->data.tid), NULL, &thread_func, &(new_node->data));
+        if( ret != 0 )
+        {
+            syslog(LOG_ERR, "pthread_create");
+        }
+        else
+        {
+            syslog(LOG_DEBUG, "pthread_create");
+        }
+
+        sll_insert(&head, new_node);
+    }
+    current =  head;
+    prev = head;
+    while(current)
+    {
+        if((current->data.complete_flag == true) && (current == head))
+        {
+     
+            head = current->next;
+            pthread_join(current->data.tid, NULL);
+            free(current);
+            current = head;
+        }
+        else if ((current->data.complete_flag == true) && (current != head)) 
+        { 
+   
+            prev->next = current->next;
+            current->next = NULL;
+            pthread_join(current->data.tid, NULL);
+            free(current);
+            current = prev->next;
+        } 
+        else 
+        {
+       
+            prev = current;
+            current = current->next;
+        }
+    }
+    printf("All threads exited!!\n");
```

Comment:
@anuh7 what about threads which haven't completed?  I think you need to wait for all threads above, not just ones which have completed.